File: core/src/main/java/com/vzome/core/editor/Polytope4d.java

//(c) Copyright 2005, Scott Vorthmann.  All rights reserved.

package com.vzome.core.editor;

import java.util.HashMap;
import java.util.Map;

import org.w3c.dom.Element;

import com.vzome.core.algebra.AlgebraicField;
import com.vzome.core.algebra.AlgebraicNumber;
import com.vzome.core.algebra.AlgebraicVector;
import com.vzome.core.commands.AttributeMap;
import com.vzome.core.commands.XmlSaveFormat;
import com.vzome.core.construction.FreePoint;
import com.vzome.core.construction.Point;
import com.vzome.core.construction.Segment;
import com.vzome.core.construction.SegmentJoiningPoints;
import com.vzome.core.math.DomUtils;
import com.vzome.core.math.Projection;
import com.vzome.core.math.QuaternionProjection;
import com.vzome.core.math.symmetry.WythoffConstruction;
import com.vzome.core.model.RealizedModel;

public class Polytope4d extends ChangeManifestations
{
    private int index;
    private final AlgebraicField field;
    private Projection proj;
    private AlgebraicVector quaternion;
    private String groupName;

    // new controls
    private int edgesToRender = 0xF;
    private AlgebraicNumber[] edgeScales = new AlgebraicNumber[4];
    private String renderGroupName;
    private final FieldApplication fieldApp;

    public Polytope4d( Selection selection, RealizedModel realized,
            FieldApplication fieldApp,
            AlgebraicVector quaternion, int index, String groupName,
            int edgesToRender, AlgebraicNumber[] edgeScales,
            String renderGroupName, boolean groupInSelection )
    {
        super( selection, realized, groupInSelection );
        this.fieldApp = fieldApp;
        this.field = fieldApp .getField();

        this.index = index;
        this.quaternion = quaternion;
        this.groupName = groupName;

        this .renderGroupName = renderGroupName;
        this .edgesToRender = edgesToRender;
        if ( edgeScales != null )
            this .edgeScales = edgeScales;
        else
            for (int i = 0; i < this .edgeScales .length; i++)
            {
                this .edgeScales[ i ] = this .field .createPower( 0 );
            }
    }

    public Polytope4d( Selection selection, RealizedModel realized,
            FieldApplication fieldApp,
            Segment symmAxis, int index, String groupName, boolean groupInSelection )
    {
        this( selection, realized, fieldApp,
                ( symmAxis == null )? null : symmAxis .getOffset() .inflateTo4d(),
                        index, groupName, index, null, groupName, groupInSelection );
    }

    @Override
    protected String getXmlElementName()
    {
        return "Polytope4d";
    }

    @Override
    public void getXmlAttributes( Element xml )
    {
        if ( quaternion != null )
            DomUtils .addAttribute( xml, "quaternion", quaternion .toParsableString() );        
        DomUtils .addAttribute( xml, "group", this.groupName );
        DomUtils .addAttribute( xml, "wythoff", Integer .toString( this.index, 2 ) );
        if ( this .edgesToRender != 0xF )
            DomUtils .addAttribute( xml, "renderEdges", Integer .toString( this.edgesToRender, 2 ) );
        if ( ! this .renderGroupName .equals( this .groupName ) )
            DomUtils .addAttribute( xml, "renderGroup", this.renderGroupName );
    }

    @Override
    public void setXmlAttributes( Element xml, XmlSaveFormat format )
    {
        String binary = xml .getAttribute( "wythoff" );
        this .index = Integer .parseInt( binary, 2 );
        String renderString = xml .getAttribute( "renderEdges" );
        this .edgesToRender = ( renderString==null || renderString .isEmpty() )? this .index : Integer .parseInt( renderString, 2 );
        this .groupName = xml .getAttribute( "group" );
        String rgString = xml .getAttribute( "renderGroup" );
        this .renderGroupName = ( rgString==null || rgString .isEmpty() )? this .groupName : rgString;

        String quatString = xml .getAttribute( "quaternion" );
        if ( quatString != null && ! "" .equals( quatString ) ) {
            // newest format
            if ( quatString .contains( "+" ) ) {
                // First, deal with the bug we had in serialization...
                quatString = quatString .replace( ',', ' ' );
                quatString = quatString .replace( '(', ' ' );
                quatString = quatString .replace( ')', ' ' );
                quatString = quatString .replace( '+', ' ' );
                char irrat = this .field .getIrrational( 0 ) .charAt( 0 );
                quatString = quatString .replace( irrat, ' ' );
                quatString = quatString + " 0 0 0";  // This is probably OK, but
                //  it is known to work only for those particular files I have seen,
                //  in which the X, Y, and Z coordinates are all zero.
            }
            this .quaternion = this .field .parseVector( quatString );
        }
        else {
            // legacy formats
            Segment segment = null;
            if ( format .commandEditsCompacted() )
                segment = format .parseSegment( xml, "start", "end" );
            else
            {
                AttributeMap attrs = format .loadCommandAttributes( xml );
                segment = (Segment) attrs .get( "rotation" );
            }
            if ( segment != null )
                this.quaternion = segment .getOffset() .inflateTo4d();
        }
    }


    @Override
    public void perform()
    {
        if ( quaternion == null )
            this.proj = new Projection .Default( field );
        else
            this.proj = new QuaternionProjection( field, null, quaternion .scale( field .createPower( -5 ) ) );
        this .fieldApp .constructPolytope( groupName, this.index, this .edgesToRender, this .edgeScales, new WythoffListener() );
        redo();
    }


    private class WythoffListener implements WythoffConstruction.Listener
    {
        private int numVertices = 0;
        Map<AlgebraicVector, Point> vertices = new HashMap<>();

        @Override
        public Object addEdge( Object p1, Object p2 )
        {
            Segment edge = new SegmentJoiningPoints( (Point) p1, (Point) p2 );
            manifestConstruction( edge );
            return edge;
        }

        @Override
        public Object addFace( Object[] vertices )
        {
            return null;
        }

        @Override
        public Object addVertex( AlgebraicVector vertex )
        {
            Point p = vertices .get( vertex );
            if ( p == null )
            {
                AlgebraicVector projected = vertex;
                if ( proj != null )
                    projected = proj .projectImage( vertex, true );

                projected = projected .scale( field .createPower( 5 ) );

                p = new FreePoint( projected );
                p .setIndex( numVertices++ );
                manifestConstruction( p );
                vertices .put( vertex, p );
            }
            return p;
        }
    }

    public static String[] getSupportedGroups()
    {
        return new String[]{ "A4", "B4/C4", "D4", "F4", "H4" };
    }
}

File: online/src/worker/legacy/ts/com/vzome/core/edits/Polytope4d.ts
import { java, javaemul } from "../../../../../candies/j4ts-2.1.0-SNAPSHOT/bundle.js";
import { AlgebraicField } from "../algebra/AlgebraicField.js";
import { AlgebraicNumber } from "../algebra/AlgebraicNumber.js";
import { AlgebraicVector } from "../algebra/AlgebraicVector.js";
import { AttributeMap } from "../commands/AttributeMap.js";
import { XmlSaveFormat } from "../commands/XmlSaveFormat.js";
import { FreePoint } from "../construction/FreePoint.js";
import { Point } from "../construction/Point.js";
import { Segment } from "../construction/Segment.js";
import { SegmentJoiningPoints } from "../construction/SegmentJoiningPoints.js";
import { ChangeManifestations } from "../editor/api/ChangeManifestations.js";
import { EditorModel } from "../editor/api/EditorModel.js";
import { SymmetryAware } from "../editor/api/SymmetryAware.js";
import { Projection } from "../math/Projection.js";
import { QuaternionProjection } from "../math/QuaternionProjection.js";
import { Symmetries4D } from "../math/symmetry/Symmetries4D.js";
import { WythoffConstruction } from "../math/symmetry/WythoffConstruction.js";
import { DomUtils } from "../../xml/DomUtils.js";
import { Element } from "../../../../org/w3c/dom/Element.js";

export class Polytope4d extends ChangeManifestations {
    /*private*/ index: number;

    /*private*/ field: AlgebraicField;

    /*private*/ proj: Projection;

    /*private*/ quaternion: AlgebraicVector;

    /*private*/ groupName: string;

    /*private*/ edgesToRender: number;

    /*private*/ edgeScales: AlgebraicNumber[];

    /*private*/ renderGroupName: string;

    /*private*/ symmetries: Symmetries4D;

    public constructor(editor: EditorModel) {
        super(editor);
        if (this.index === undefined) { this.index = 0; }
        if (this.field === undefined) { this.field = null; }
        if (this.proj === undefined) { this.proj = null; }
        if (this.quaternion === undefined) { this.quaternion = null; }
        if (this.groupName === undefined) { this.groupName = null; }
        this.edgesToRender = 15;
        this.edgeScales = [null, null, null, null];
        if (this.renderGroupName === undefined) { this.renderGroupName = null; }
        if (this.symmetries === undefined) { this.symmetries = null; }
        this.symmetries = (<SymmetryAware><any>editor).get4dSymmetries();
        this.field = editor.getRealizedModel().getField();
        for(let i: number = 0; i < this.edgeScales.length; i++) {{
            this.edgeScales[i] = this.field.one();
        };}
    }

    /**
     * 
     * @param {*} params
     */
    public configure(params: java.util.Map<string, any>) {
        this.groupName = <string>params.get("groupName");
        this.renderGroupName = <string>params.get("renderGroupName");
        this.index = (<number>params.get("index")|0);
        this.edgesToRender = (<number>params.get("edgesToRender")|0);
        this.edgeScales = <AlgebraicNumber[]>params.get("edgeScales");
        this.quaternion = <AlgebraicVector>params.get("quaternion");
    }

    /**
     * 
     * @return {string}
     */
    getXmlElementName(): string {
        return "Polytope4d";
    }

    /**
     * 
     * @param {*} xml
     */
    public getXmlAttributes(xml: Element) {
        if (this.quaternion != null)DomUtils.addAttribute(xml, "quaternion", this.quaternion.toParsableString());
        DomUtils.addAttribute(xml, "group", this.groupName);
        DomUtils.addAttribute(xml, "wythoff", DomUtils.byteToBinary(this.index));
        if (this.edgesToRender !== 15)DomUtils.addAttribute(xml, "renderEdges", DomUtils.byteToBinary(this.edgesToRender));
        if (!(this.renderGroupName === this.groupName))DomUtils.addAttribute(xml, "renderGroup", this.renderGroupName);
    }

    /**
     * 
     * @param {*} xml
     * @param {XmlSaveFormat} format
     */
    public setXmlAttributes(xml: Element, format: XmlSaveFormat) {
        const binary: string = xml.getAttribute("wythoff");
        this.index = javaemul.internal.IntegerHelper.parseInt(binary, 2);
        const renderString: string = xml.getAttribute("renderEdges");
        this.edgesToRender = (renderString == null || /* isEmpty */(renderString.length === 0)) ? this.index : javaemul.internal.IntegerHelper.parseInt(renderString, 2);
        this.groupName = xml.getAttribute("group");
        const rgString: string = xml.getAttribute("renderGroup");
        this.renderGroupName = (rgString == null || /* isEmpty */(rgString.length === 0)) ? this.groupName : rgString;
        let quatString: string = xml.getAttribute("quaternion");
        if (quatString != null && !("" === quatString)){
            if (/* contains */(quatString.indexOf("+") != -1)){
                quatString = /* replace */quatString.split(',').join(' ');
                quatString = /* replace */quatString.split('(').join(' ');
                quatString = /* replace */quatString.split(')').join(' ');
                quatString = /* replace */quatString.split('+').join(' ');
                const irrat: string = this.field['getIrrational$int'](0).charAt(0);
                quatString = /* replace */quatString.split(irrat).join(' ');
                quatString = quatString + " 0 0 0";
            }
            this.quaternion = this.field.parseVector(quatString);
        } else {
            let segment: Segment = null;
            if (format.commandEditsCompacted())segment = format.parseSegment$org_w3c_dom_Element$java_lang_String$java_lang_String(xml, "start", "end"); else {
                const attrs: AttributeMap = format.loadCommandAttributes$org_w3c_dom_Element(xml);
                segment = <Segment>attrs.get("rotation");
            }
            if (segment != null)this.quaternion = segment.getOffset().inflateTo4d$();
        }
    }

    /**
     * 
     */
    public perform() {
        if (this.quaternion == null)this.proj = new Projection.Default(this.field); else this.proj = new QuaternionProjection(this.field, null, this.quaternion.scale(this.field['createPower$int'](-5)));
        this.symmetries.constructPolytope(this.groupName, this.index, this.edgesToRender, this.edgeScales, new Polytope4d.WythoffListener(this));
        this.redo();
    }

    public static getSupportedGroups(): string[] {
        return ["A4", "B4/C4", "D4", "F4", "H4"];
    }
}
Polytope4d["__class"] = "com.vzome.core.edits.Polytope4d";


export namespace Polytope4d {

    export class WythoffListener implements WythoffConstruction.Listener {
        public __parent: any;
        numVertices: number;

        vertices: java.util.Map<string, Point>;

        /**
         * 
         * @param {*} p1
         * @param {*} p2
         * @return {*}
         */
        public addEdge(p1: any, p2: any): any {
            const edge: Segment = new SegmentJoiningPoints(<Point>p1, <Point>p2);
            this.__parent.manifestConstruction(edge);
            return edge;
        }

        /**
         * 
         * @param {java.lang.Object[]} vertices
         * @return {*}
         */
        public addFace(vertices: any[]): any {
            return null;
        }

        /**
         * 
         * @param {AlgebraicVector} vertex
         * @return {*}
         */
        public addVertex(vertex: AlgebraicVector): any {
            let p: Point = this.vertices.get(vertex.toString());
            if (p == null){
                let projected: AlgebraicVector = vertex;
                if (this.__parent.proj != null)projected = this.__parent.proj.projectImage(vertex, true);
                projected = projected.scale(this.__parent.field['createPower$int'](5));
                p = new FreePoint(projected);
                p.setIndex(this.numVertices++);
                this.__parent.manifestConstruction(p);
                this.vertices.put(vertex.toString(), p);
            }
            return p;
        }

        constructor(__parent: any) {
            this.__parent = __parent;
            this.numVertices = 0;
            this.vertices = <any>(new java.util.HashMap<any, any>());
        }
    }
    WythoffListener["__class"] = "com.vzome.core.edits.Polytope4d.WythoffListener";
    WythoffListener["__interfaces"] = ["com.vzome.core.math.symmetry.WythoffConstruction.Listener"];


}
